### PR TITLE
[Matrix-Generation] Support additional customized steps before matrix generation

### PR DESCRIPTION
**Pull request by scbedd**
@benbp 

This extension will allow the python regression testing to dynamically update it's matrix.

Related to Azure/azure-sdk-for-python#23519

### CODE REVIEW

File: eng/common/scripts/job-matrix/README.md [hunk 1]
```diff
@@ -56,6 +56,7 @@ jobs:
         Cloud: Public
       MatrixFilters: []
       MatrixReplace: []
+      PreGenerationSteps: []
```
Comment (scbedd) [inline eng/common/scripts/job-matrix/README.md:59]: I can put a simple

> Write-Host gci $(ClonedRepo) 

Or something if we want to have an example of this in the methodolgoy. We don't have that for the "further" features like MatrixFilters or MatrixReplace though, so I feel like we're probably fine.

**Review by benbp — APPROVED** [review]

File: eng/common/scripts/job-matrix/README.md (hunk 1 above)
Comment (benbp) [reply eng/common/scripts/job-matrix/README.md:59]: Or a section in the docs regarding dynamic matrix generation in a pipeline: https://github.com/Azure/azure-sdk-tools/blob/main/eng/common/scripts/job-matrix/README.md#usage-in-a-pipeline

**Comment by scbedd** [comment]
Well I was going to post a comment with a "all done" confirming that this works...and it doesn't.


[Build](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1433372&view=results)
[Commit](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1433372&view=results)

Currently debugging before continuing with the common update.

File: eng/common/scripts/job-matrix/README.md (hunk 1 above)
Comment (scbedd) [reply eng/common/scripts/job-matrix/README.md:59]: My paranoia panned out. This doesn't work.

**Comment by scbedd** [comment]
After steplist update, [checks out.](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1435987&view=logs&j=76992fd1-2312-5fae-7c45-7baba86b87ae&t=e370f21b-6146-5cbb-27bc-c744450b188c)

Advancing this to merge.

File: eng/common/scripts/job-matrix/README.md (hunk 1 above)
Comment (scbedd) [reply eng/common/scripts/job-matrix/README.md:61]: @benbp added a readme section as requested.

**Comment by scbedd** [comment]
I want to squash this set of commits down, but have already pushed the results (and squashed) to each destination repo. Going to check-enforcer override, then merge this build manually to get the commit history I want.

Next time I'll avoid the situation entirely.

**Pull request merged**